Revive beerocks analyzer

rmelotte opened the pull request:
For the moment, the focus is on the connectivity map part

rmelotte committed:
tools: beerocks_analyzer: switch to PySide2

PySide is deprecated and does not support Python versions greater than 3.4.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: update each requirement to its latest version

Also update matplotlib's imports.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: set minimum Python version to 3.5

Python 2 will soon be deprecated and some dependencies, like
"networkx", are not compatible with Python 2 anymore.

This commit sets the minimum Python version to 3.5.
The shebang has also been updated to use python3 instead of python,
and to look for it in the environment instead of always using /usr/bin/python3.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: import QtWidgets

QMainWindow has moved to QtWidgets.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: switch xrange usage to range

xrange is not available anymore in Python3, but Python3's range is
similar to Python2 xrange and can be used in-place.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: use a logger instead of PRINTF

Using a logger gives more control over what gets output to the
console, to a file, etc.

Remove the definitions of PRINTF.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: remove the custom package installation

The custom package installation install_required_pkg is unreliable and
hard to maintain. Instead, rely on the simple `pip install -r
requirements.txt` (which will be documented in the README.md) to
install the required packages.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: fix mac conversion

There is no need to convert to and from hex to insert ':' in-between.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: sleep between two file reads

Nothing has changed except sleeping for 1 second if the line is
empty, and adding two more logging calls.

This prevents the reading thread from hogging the cpu if there is
nothing to read.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

morantr reviewed (dismissed):
ed2f731e8c68b06d90147d9d7d5d59114ee77f9e commit message headline length is greater than 72 chars, and wrapped.
f6ca385f5c15acc40a3e56260c74f1d411a4d865 commit message headline typo: usr -> use
23c2d9e48d308d98715b94f4c0382a2a41f75e13 commit message headline length is greater than 72 chars, and wrapped.

Beside that,
1. The component part of the commit message headline is not supposed to be a file path.
Please change it to `tools: beerocks_analyzer: ` on all commits.

2. Most of the commits are not descriptives enough, and missing:
    - the context (current situation);
    - the problem statement (why the current situation is not good);
    - possible solutions with their pros and cons;
    - the chosen solution and why it was chosen;

    (https://github.com/prplfoundation/prplMesh/blob/master/CONTRIBUTING.md#commits)
    This would be more easy to review and understand, especially for a python beginner like me.

3. I see that on beerocks analyzer folder there is a reuierments.txt file. It does not have updated versions anymore after you updated some of the python packages.
I think it would be a great idea to replace this file with a markdown file, which says which packages are needed, and explains how to use the beerocks analyzer.

rmelotte commented:
Thanks for the comments! I'll improve the commits messages.

> I see that on beerocks analyzer folder there is a reuierments.txt file. It does not have updated versions anymore after you updated some of the python packages.
I think it would be a great idea to replace this file with a markdown file, which says which packages are needed, and explains how to use the beerocks analyzer.

I don't really understand the second sentence. Could you rephrase it ?
The requirements.txt file is needed for pip to be able to install the required dependencies.
I will add a README or even a bash script to run the tool once it's working.

rmelotte committed:
tools: beerocks_analyzer: remove unused node attribute

The "Position" argument was added to the node, but it was never
used.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: fix deprecated networkx node usage

graph.node has been replaced by graph.nodes and will be removed in the
future.

See networkx/classes/graph.py.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: change ap_edges colors

Previosuly, edges going to/coming from ETH nodes where white (so not
visible on the white background).

It's actually useful to see those edges so change their color to gray.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: connect nodes to the gateway

We want to visually see the IRE connected to the gateway.

Previously, only nodes which had a parent_bssid where connected.

This commit connects every node that doesn't have a parent_bssid to
the gateway by setting it's parent_mac to the gateway's parent_mac.

Note that this is probably not the way it was supposed to work
originally and it's just a quick fix to have the visual result we want
until a better solution is implemented.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: fix dict key availability checking

dict.has_keys() is no longer in Python3.

The Pythonic way to do it in Python3 (and 2) is to simply use "in" and
the dictionary itself.

This commit fixes has_keys() usage, and also the drops the old '%'
operator usage for string formatting in favour of the new .format().

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: fix '>' comparison against None

Python3 does not allow '>' comparisons against None anymore, so None
values have to be checked before doing the comparison.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: do not print traces on missing info

Previously the logger would print a stacktrace when the bandwidth or
channel was missing.

A missing bandwidth or channel is not a big issue, so this commit
only logs it as at the debug level.

While we're at it, also narrow down the exception to ValueError.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: fix socket timeout handling

beerocks_analyzer has a socket to receive data from beerocks_cli.

When there was no data read, it should not try to use
the (non-existent) data.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: clear the graph when receiving the mark

beerocks_analyzer uses a mark to know when the connection map is
complete.
Previously, when a node disappeared from one update to the
other (without receiving a Disconnect event), it was kept in the graph.

This commit clears the graph when a new mark is read, to make sure
that the graph represents the new state of the map.

The visual effect is not so great, but that will be fixed in a later
commit.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: don't rebuild the entire graph each time

As the analyzer does not know when it received a full nw_map update or
only part of it, it cannot immediately remove nodes that disappear from
one update to the other.

The way it was dealt with before was to rebuild the whole nw_map each
time, which was not visually comfortable.

This commit keeps track of when the nodes where last seen, and removes
them if they have not been seen since a certain amount of updates.

This has the downside that it makes disconnected nodes disappear only
after some time, but counters the fact that there is currently no
event received when a node disconnected, and that the analyzer does
not know when the update has been fully received.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

morantr commented:
1. https://github.com/prplfoundation/prplMesh/pull/308/commits/7069fe0dd59f5b64202d7cc8685228da40d3c88c and https://github.com/prplfoundation/prplMesh/pull/308/commits/e366b8f890fd19570c627d644b00e73d689e4069 have the same headline (no description tough). Maybe can you unify them?

2. Same thing for https://github.com/prplfoundation/prplMesh/pull/308/commits/0f6dc748a5a64660a4a497b292495dc8a1b896a1 and https://github.com/prplfoundation/prplMesh/pull/308/commits/b501ecabc5646fbc6568735ef81521ef7af224c9. Both of them have a headline which is pretty similar which is very generic. Please be more specific about what you change. If there isn't much difference squash them.

3. https://github.com/prplfoundation/prplMesh/pull/308/commits/1a84e2000ca54cac25feccb18134fb930f4e7049 has too generic headline (no description). Please write what you fix explicitly.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -40,29 +40,26 @@
     global t_list, g_run_flag
 
-    PRINTF("")
+    logger.info("Signal received: " + str(signal))
```
It is better to make a big change like this purely mechanical (i.e. replace all occurrences of PRINTF with logger.info), and do other changes (change info to something else, add more info to the log, add a TODO comment) in separate commits.

If you really do it mechanically (e.g. with a sed program), then include the scriptlet you used in the commit message, and explicitly mention any manual fixups you needed to do afterwards.

However, it's not worth spending effort now to try to split up this commit. Just something to remember for the future.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -176,15 +173,15 @@ def openLogFiles(self, log_start_pos=0):
                 self.fdLog = open(self.fname,"rt")
                 if log_start_pos > 0:
-                    # PRINTF("openLogFiles: seek to %d\n" % log_start_pos)
+                    logger.debug("openLogFiles: seek to {:d}", log_start_pos)
```
About the "purely mechanical": this is one which would make sense to still do in the same commit, but then the commit message should mention this explicitly, e.g. "Replace commented-out PRINTF calls with logger.debug. Replace `%`-formatted strings with the `{}`-formatting used by logger."

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -685,8 +685,8 @@ def socket_server(log_file):
                 try:
-                    i=data.index('\n')
                 except:
+                    i=data.index(r'\n')
```
This makes no sense - a try-block without body is meaningless...

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -706,13 +710,42 @@ def socket_server(log_file):
+
+    command = str(beerocks_cli_path) + " -a " + str(target_ip)
+    logger.debug("Starting beerock_cli -a")
+    exit_status = subprocess.call(command.split())
```
There shouldn't be a reason to build up the path as a string and then to split it. I.e., use:

```
command = [str(beerocks_cli_path), "-a", str(target_ip)]
exit_status = subprocess.call(command)
```

And in this case, I don't even think the explicit conversions to str are needed.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -706,13 +710,42 @@ def socket_server(log_file):
     file.close()
 
+
+def local_start_beerocks_cli(beerocks_cli_path: Path, target_ip: IPv4Address) -> int:
```
Type info, nice!

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -676,11 +676,13 @@ def socket_server(log_file):
+                data += connection.recv(256).decode("utf-8")
             except socket.timeout:
-                logger.info("Socket timed out")
+                logger.debug("Socket timed out")
```
This change seems to be unrelated/unneeded.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -340,43 +340,47 @@ def readSampleThread(self, update_start_time = False):
-                                    self.cm_widget.readSample(line, param_t1, is_start, is_stop)
-                                if self.isGraphs:
-                                    self.wa_widget.readSampleAndUpdateGraphs(line, param_t1, is_start, is_stop)
+                if len(line) <= 0:
```
`len(line)` can never be smaller than 0, so this should be `len(line) == 0`, but then it's better to just do `if not line:`

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -340,43 +340,47 @@ def readSampleThread(self, update_start_time = False):
+                if len(sta_ap_ids) > 0:
+                    line += ", " + sta_ap_ids
+                if is_stop:
+                    logger.info("Read stop flag, stopping")
```
It is ***really bad*** to hide additional changes in a big indentation change...

arnout commented on tools/beerocks_analyzer/beerocks_analyzer_widget.py:
```diff
@@ -19,8 +19,7 @@
 
-matplotlib.use('Qt4Agg')
-matplotlib.rcParams['backend.qt4']='PySide'
+matplotlib.use('Qt5Agg')
```
The same change was made in a different file two commits before. It's better to do those two changes together.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -734,9 +738,11 @@ def local_start_beerocks_cli(beerocks_cli_path: Path, target_ip: IPv4Address) ->
-    logger.debug("Starting beerock_cli -a")
+    command = ""
+    if docker_container_name:
+        command += "docker exec " + docker_container_name + " "
```
It makes more sense to me to first build the basic command and then prepend the rest. With the list, that becomes:

```python
command = [str(beerocks_cli_path), "-a", str(target_ip)]
if docker_container_name:
    command = ["docker", "exec", docker_container_name] + command
```

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -57,18 +57,15 @@ def printUsage():
     global t_list, g_run_flag
 
-    logger.info("Signal received: " + str(signal))
+    logger.info("Signal {} received \n terminating".format(str(signal)))
```
No need for `.format`, logger.info already implies it.

arnout reviewed (approved):
I have a bunch of comments on the commit messages.

- Summary line should never end with a period (cfr. CONTRIBUTING.md).
- "fix the python version check" is not a good commit message; you can make it more specific, like "set minimum python version to 3.5". The commit message should also explain why you need this. The next commit message ("import QtWidgets") is a good example of the right way to do it.
- Similar comment for "fix xrange usage".
- Also "fix mac conversion" warrants a better explanation of why it is wrong.
- Also "fix old Python2 usage" is not a good summary line.
- In that same commit message, you use "directory" to mean "dictionary".
- A few commits later, "fix Python2 usage" again...
- "implement threads termination" is a bit too big for easy review. It should have been split into a commit that introduces the SocketServerThread class (which is largely an indentation change), one that introduces BeerocksCliThread, etc. and finally one that implements terminate on each of them.

Since all of the comments are of the nitpick kind, I'm approving already. So please fix as much as you can/like/have time for, rebase on master and force-push. And set this PR to ready of course.

rmelotte committed:
tools: beerocks_analyzer: revive the SSH option

It was previously possible to connect to the controller via SSH, but
it did not work anymore.

This commit allows to usage the SSH option again.
It removes the usage of a password, which means that you need a
private key to be registered in your local ssh agent.
A future commit will explain how to do that.

Note that the current version of paramiko is not compatible with the
latest openssh private key format, so a temporary fork is
needed (included in the requirement.txt).

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: add a .gitignore

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: add a README

Also add a top-level .gitignore for the venv directory, and link the
analyzer's README in the top-level README.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>
Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

rmelotte commented:
Thanks for all the comments, I still need to do that quickly before marking it ready for review.
I just pushed what I have now anyway so that anybody wanting to use it with SSH can still do it already.

rmelotte committed:
tools: beerocks_analyzer: log exceptions instead of silencing them

Exceptions were previously caught but nothing was done with them, so
they were only silenced. This commit uses the logger to at least log
the exceptions, so that we know they happened.

Exceptions still have to be handled in a cleaner way by only catching
specific ones and not the broader Exception, and taking
appropriate actions based on each specific case.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: fix networkx graph.successors usage

graph.successors now returns an iterator, so it's not possible to have
its length without constructing a whole list.

In this case, out_degree() give us the information we need, so use
that instead.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: decode data from socket

With Python 3, data from the socket needs to be decoded before it
can be used as a string.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte commented:
I believe I addressed most of the issues, except for the "implement threads termination" commit, which would require a bit more time to be properly split in separate commits.
I'm marking the PR to ready, feel free to comment if you think that commit should be split anyway.

arnout commented on tools/beerocks_analyzer/connectivity_map_widget.py:
```diff
@@ -195,11 +195,11 @@ def add_delta_x_from_node(self, n, delta):
     def update_x_positions_from_node(self, n):
         #check if n has clients connected
-        num_of_successors =len(self.graph.successors(n))
+        num_of_successors =len(list(self.graph.successors(n)))
```
The proper way to do this is actually `num_of_successors = self.graph.out_degree(n)`.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer_widget.py:
```diff
@@ -19,8 +19,7 @@
-matplotlib.rcParams['backend.qt4']='PySide'
+matplotlib.use('Qt5Agg')
 import matplotlib.pyplot as plt
 from matplotlib.backends.backend_qt4agg import FigureCanvasQTAgg as FigureCanvas
```
Shouldn't this be changed into qt5 as well? It does work because the qt4 backend just forwards to the qt5 backend, but we can expect that to be removed from matplotlib at some point...

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -734,9 +738,11 @@ def local_start_beerocks_cli(beerocks_cli_path: Path, target_ip: IPv4Address) ->
+    if docker_container_name:
+        command += "docker exec " + docker_container_name + " "
+    command += str(beerocks_cli_path) + " -a " + str(target_ip)
+    logger.debug(command)
```
The original debug message was a bit better, so I'd make this `logger.debug("Starting '%s'", ' '.join(command))` (taking into account that I think command should be a list instead of a string).

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -870,8 +898,7 @@ def main(argv):
         for t in t_list:
-            logger.debug("Joining thread {}" + str(t))
-            t.join()
+            t.terminate()
```
This could be factored with the KeyboardInterrupt, i.e. just do the terminate after the except clause.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -632,147 +629,176 @@ def show():
+    def __init__(self, log_file):
+        super().__init__()
+        self.log_file = log_file
+        self.run_local = True
```
Since all paths that set g_run_flag also call this thread's terminate function, you can in fact eliminate g_run_flag.

Also, I think run_flag or continue_running or something like that is more appropriate than run_local. run_local doesn't tell me what it's supposed to do.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -632,147 +629,176 @@ def show():
-            while g_run_flag:
+class SocketServerThread(threading.Thread):
+
+    def __init__(self, log_file):
```
In other commits you added typeinfo to new functions. Would be nice to do that here as well.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -624,129 +629,203 @@ def show():
+
+    def terminate(self):
+        self.logger.debug("{} terminating".format(self.__class__.__name__))
+        self.process.kill()
```
It would make sense to add a `if self.process` here, in case terminate gets called before the subprocess is even started. But that brings us to the annoying limitation in Python that there's still a race between the creation of the Popen object and its assignment to self.process...

rmelotte replied:
Do you have a link explaining that limitation ?
I added the `if self.process` condition, but I would like to add a note about the race condition.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -624,129 +629,203 @@ def show():
+        self.ssh_client = paramiko.SSHClient()
+
+    def run(self):
+        self.ssh_client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
```
Since paramiko is not that well maintained, maybe it would be a good idea to switch to pxssh in pexpect. But that's not really necessary for this series.

rmelotte replied:
We're using a temporary fork of paramiko that is not maintained, but once [this PR](https://github.com/paramiko/paramiko/pull/1343) is merged, we will be able to switch to upstream paramiko again (which is still maintained as of today).

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -624,129 +629,203 @@ def show():
+        in_buff = ''
+        while not in_buff.endswith(':~# '):
+            in_buff += ssh_shell.recv(9999).decode("utf-8")
+            if in_buff.endswith('password: '):
```
If support for passwords is removed, this condition can be removed as well.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -624,129 +629,203 @@ def show():
+            if in_buff.endswith('password: '):
+                time.sleep(3)
+                in_buff = ''
+            elif 'are you sure you want to continue connecting' in in_buff:
```
With pxssh, this is handled automatically.

rmelotte replied:
After looking at it, there is actually a way to handle this in paramiko by either auto-adding the key or raising a warning. I went for the warning.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -624,129 +629,203 @@ def show():
+        self.ssh_client.connect(self.host, port=self.ssh_port, username=self.user)
+
+        if not self.my_ip:
+            self.my_ip = self.ssh_client.get_transport().sock.getsockname()[0]
```
Hm, pxssh unfortunately doesn't have an equivalent for this... But you could use the ssh options to create a tunnel, so that you'd always tell beerocks_cli to connect to a local port.

But then, it would probably be better to interact directly over the ssh connection instead of creating an additional connection.

rmelotte replied:
Even without using pxssh, avoiding the additional connection would be an improvement, I'll open an issue about that.

arnout commented on tools/beerocks_analyzer/.gitignore:
```diff
@@ -0,0 +1,272 @@
+
+# Created by https://www.gitignore.io/api/vim,emacs,python,pycharm
+# Edit at https://www.gitignore.io/?templates=vim,emacs,python,pycharm
```
That website is *evil*. Or you're not using it correctly. Quoting the gitignore man page:

> Which file to place a pattern in depends on how the pattern is meant to be used.
> 
> - Patterns which should be version-controlled and distributed to other repositories via clone (i.e., files that all developers will want to ignore)  should go into a .gitignore file.
> - Patterns which are specific to a particular repository but which do not need to be shared with other related repositories (e.g., auxiliary files that live inside the repository but are specific to one user’s workflow) should go into the $GIT_DIR/info/exclude file.
>  - Patterns which a user wants Git to ignore in all situations (e.g., backup or temporary files generated by the user’s editor of choice) generally go into a file specified by core.excludesFile in the user’s ~/.gitconfig. Its default value is $XDG_CONFIG_HOME/git/ignore. If $XDG_CONFIG_HOME is either not set or empty, $HOME/.config/git/ignore is used instead.

I think the only ignore pattern that is relevant here is `__pycache__/`. And that pattern should go to tools/.gitignore, which already has `*.pyc`. Well, actually, most of the gitignore files should be hoisted up to the top-level, but that's a different issue :-)

arnout commented on tools/beerocks_analyzer/README.md:
```diff
@@ -0,0 +1,77 @@
+Warning: there still is an enormous list of issues with `beerocks_analyzer`, and only the connection map currently works. This README will walk you through installing dependencies and displaying the connection map.
```
Please also update the top-level README.md to direct to this specific documentation.

Also, please follow the coding style for markdown files: one line per sentence.

Finally, please add a H1 header to act as a title:

```markdown
# beerocks analyzer
```

arnout commented on tools/beerocks_analyzer/README.md:
```diff
@@ -0,0 +1,77 @@
+Warning: there still is an enormous list of issues with `beerocks_analyzer`, and only the connection map currently works. This README will walk you through installing dependencies and displaying the connection map.
+
+
+# Prerequisites
```
Use H2 (`##`) for sectioning.

Hm, maybe we should add that to the coding style...

arnout commented on tools/beerocks_analyzer/README.md:
````diff
@@ -0,0 +1,77 @@
+## Create a virtualenv to use the application
+
+```sh
+python3 -m venv venv
````
Ah, this explains why the venv stuff was added to .gitignore. But you may want to call venv from the top-level directory, so it should be added to the top-level .gitignore.

However, wouldn't it be better to give the venv a more specific name, e.g. `analyzer`?

arnout commented on tools/beerocks_analyzer/README.md:
````diff
@@ -0,0 +1,77 @@
+You need to give the path to beerock's binaries (change `BIN_PATH` according to your installation):
+
+```sh
+BIN_PATH="/home/$USER/prplmesh_root/build/install/bin/beerocks_cli"
````
Actually, we assume in dozens of other places that the install directory is `../build/install` compared to the top-level of prplMesh, so we could do the same in beerocks_analyzer.py. I.d., calculate the default bin_path based on the location of the script:
```python
script_dir = os.path.dirname(os.path.abspath(sys.argv[0]))
top_dir = os.path.abspath(os.path.join(script_dir, '..', '..'))
bin_path = os.path.abspath(os.path.join(top_dir, '..', 'build', 'install', 'bin', 'beerocks_cli'))
```

And maybe the abspath and intermediate top_dir are not even needed.

arnout commented on tools/beerocks_analyzer/README.md:
````diff
@@ -0,0 +1,77 @@
+
+```sh
+BIN_PATH="/home/$USER/prplmesh_root/build/install/bin/beerocks_cli"
+python beerocks_analyzer.py -map -bin_path="$BIN_PATH" -docker_container=gateway
````
`-docker_container` is wrong here.

python is wrong, it should be python3. But if we fix the script's shebang, it's not even needed (i.e. use `/usr/bin/env python3` instead of `/usr/bin/python`).

With the above suggestion, this could be as simple as

```sh
./beerocks_analyzer.py -map`
```

arnout commented on tools/beerocks_analyzer/README.md:
````diff
@@ -0,0 +1,77 @@
+
+```sh
+BIN_PATH="/home/$USER/prplmesh_root/build/install/bin/beerocks_cli"
+MY_IP=172.19.0.1
````
You can get this address with `docker network`, like we do in some of the test scripts.

arnout commented on tools/beerocks_analyzer/README.md:
```diff
@@ -0,0 +1,77 @@
+
+You first need to have a private key (one that can connect to the remote device) registered in your local agent, so that you don't get asked for a password.
+
+If you don't have one running, you can run an agent by doing `` eval `ssh-agent -s` ``. To register your key, just do `ssh-add`.
```
ssh-agent is only needed if the private key is password protected. I would assume that people who have a password-protected ssh key will also have an agent running already.

arnout commented on tools/beerocks_analyzer/README.md:
```diff
@@ -0,0 +1,77 @@
+
+## Using a remote controller over SSH
+
+You first need to have a private key (one that can connect to the remote device) registered in your local agent, so that you don't get asked for a password.
```
Could you update the Turris Omnia wiki page with information on how to add your public key to the device?

arnout commented on tools/beerocks_analyzer/README.md:
````diff
@@ -0,0 +1,77 @@
+To run the analyzer, you then need to give the path to beerock's binaries (change `BIN_PATH` according to your installation) and the target IP:
+
+```sh
+BIN_PATH="/home/$USER/prplmesh_root/build/install/bin/beerocks_cli"
````
That doesn't seem right...

arnout reviewed (changes requested):
I still have a bunch of comments, and some of these are actually important.

I don't want to rush the inclusion of beerocks_analyzer, people can always use the branch if they need to.

The switch to pxssh is clearly something that doesn't need to be done in this PR. Could you create a feature request issue for it? You can also mark it as "good first issue".

rmelotte committed:
tools: beerocks_analyzer: implement terminate for the socket thread

This commit implements a specific terminate() method for the
socket_server thread, by subclassing the Thread class and moving the
thread function into its run() function. The thread function arguments
become arguments of the constructor.

This also gives us the possibility to completely drop the global
g_run_flag and use an attribute of SocketServerThread.

While we're at it, add type hinting for the contructor arguments, and
document the constructor.

Also set the thread to be daemon so that it's killed anyway when the
application exits.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: add a way to use a local controller

Add the -bin-path option to allow override of the default location of
beerocks_cli. This is necessary because the local build doesn't install
in /opt. If -bin-path is not given, the following location will be
tried: <prplMesh_top_level>/../build/install/bin/beerocks_cli

If -gw-ip is not given on the command line, assume a local controller.

Add a new function local_start_beerocks_cli() to start on the local
controller. This new function uses the python typeinfo feature to make
it more self-descriptive and future-safe.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: implement terminate for the ssh thread

This commit implements a specific terminate() method for the ssh thread,
by subclassing the Thread class and moving the thread function into its
run() function. The thread function arguments become arguments of the
constructor.

This also gives us the possibility to move the global g_ssh as an
attribute of SSHThread.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: call terminate on each thread

Now that a specific terminate method has been implemented for each
thread, we can call it in an attempt to terminate threads when the
application exits.

Note that there is still a problem with the way the threading is done
in the application (resources like files are not always dealt with
correctly - use context managers) but it will mitigate risks of having
background threads that keeps running after the application exits.

Also refactor the KeyboardInterrupt except clause to do the same thing
as when the application exits normally.

While we're at it, change the logger message to output the signal
number and say the application is going to be terminated when
receiving a signal.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: add a way to use a local docker container

This commit adds a command line option to be able to use the analyzer
locally, but with the controller running inside a local docker
container.

Add a command-line option "-docker_container", to specify the docker
container name. If no docker container name is given, keep the
previous behavior and assume a local (non-docker) controller.

If the my_ip option is not used:
- if docker is used, try to use the gateway of one of the docker
  networks of the container.
- if running locally, keep using localhost as a target IP.

We also need to explicitly convert beerocks_cli_path to str, because
join() only works on str. Another possibility would be to do a
join(map(str, command)), but converting beerocks_cli_path is harmless
anyway as subprocess can work with str as well as with Path.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: add a logger

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: fix imp deprecated usage

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: implement terminate for the cli thread

This commit implements a specific terminate() method for the
cli thread, by subclassing the Thread class and moving the
thread function into its run() function. The thread function arguments
become arguments of the constructor.

The call to subprocess.call has been replaced by subprocess.Popen, to
be able to kill the process when terminate() is called.

Set the thread to be daemon so that it's killed anyway when the
application exits.

Also the cli's stdin and stderr will be printed on the corresponding
logger levels.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: docker: put the runners on the same subnet as the gateway

Previously the runners were assigned '/32' IP's, so they didn't have
any route to their gateway.

This commit assigns thems '/16' IPs (same as the gateway), so that
they get a route to the gateway by default.

A new function is added to get the length of the netid, then it's used
for the container's IP address.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: beerocks_analyzer: remove ap_ids construction

Some parts of the file parser was trying to assign unique ap_ids in a
way that was way more complex than what it needs to be.

Also it relied on parent_bssid, which is not in the data anymore.

This commit removes the ap_ids, which will be reimplemented in a
cleaner way later when/if it's needed.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

arnout commented on tools/beerocks_analyzer/beerocks_analyzer_widget.py:
```diff
@@ -488,7 +504,7 @@ def addRemoveNewFigSubplots(self, ar_t, ar_v, target_lable, is_add, entity, enti
+                                self.logger.error("Error, color == None, param_n={}".format(param_n))
                                 return
                             marker = self.marker[0]
                             if self.fig_subplot_nums2marker.has_key((f,p)):
```
There are still some instances of has_key let in beerocks_analyzer_widget.py.

(nitpick)

rmelotte replied:
This could have been fixed at the same time indeed.
Just a note however, beerocks_analyzer_widget.py is not used for the connection map and still needs a its own lot of fixes.

arnout commented on tools/beerocks_analyzer/beerocks_analyzer.py:
```diff
@@ -624,129 +628,238 @@ def show():
+        command = 'LD_LIBRARY_PATH=/opt/beerocks/lib /opt/beerocks/bin/beerocks_cli -a {}'.format(self.my_ip)
+
+        #create ssh shell
+        ssh_shell = self.ssh_client.invoke_shell()
```
It's probably possible to tell paramiko to start beerocks_cli directly instead of starting the shell. But then the LD_LIBRARY_PATH issue has to be solved first. (nitpick)

arnout commented on tools/beerocks_analyzer/README.md:
```diff
@@ -0,0 +1,87 @@
+
+### `Path to beerocks_cli not found`
+
+If you get `ValueError: Path to beerocks_cli not found`, it means the analyzer couldn't find the path to beerocks<sub>cli</sub>'s binary.
```
`beerocks<sub>cli</sub>`: I guess the org -> md conversion is not perfect :-)

rmelotte replied:
The conversion was correct, but the org file wasn't ('_' is used as subscripts like with LaTeX).

arnout commented on tools/beerocks_analyzer/doc/README.org:
```diff
@@ -0,0 +1,71 @@
+#+OPTIONS: toc:nil \n:t
```
I don't think it makes a lot of sense to include a .org in the repository, since a .md works just as well and that's what we standardised on. So I removed it. Feel free to put it back if you feel it adds value (but then defend why :-).

arnout reviewed (approved):
Just nitpicks, and I fixed up a few things myself. I'm going to merge now. If anything more needs to be changed, it will have to be in follow-up commits.
